include/ginkgo/core/matrix/diagonal.hpp

```diff
@@ -0,0 +1,191 @@
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CORE_MATRIX_DIAGONAL_HPP_
+#define GKO_CORE_MATRIX_DIAGONAL_HPP_
+
+
+#include <ginkgo/core/base/lin_op.hpp>
+
+
+namespace gko {
+namespace matrix {
+
+
+template <typename ValueType, typename IndexType>
+class Csr;
+
+template <typename ValueType>
+class Dense;
+
+
+/**
+ * This class is a utility which efficiently implements the diagonal matrix (a
+ * linear operator which scales a vector row wise).
+ *
+ * Objects of the Diagonal class always represent a square matrix, and
+ * require one array to store their values.
+ *
+ * @tparam ValueType  precision of matrix elements
+ * @tparam IndexType  precision of matrix indexes of a CSR matrix the diagonal
+ *                    is applied or converted to.
+ *
+ * @ingroup diagonal
+ * @ingroup mat_formats
+ * @ingroup LinOp
+ */
+template <typename ValueType = default_precision, typename IndexType = int32>
+class Diagonal : public ConvertibleTo<Csr<ValueType, IndexType>>,
+                 public EnableLinOp<Diagonal<ValueType, IndexType>>,
+                 public EnableCreateMethod<Diagonal<ValueType, IndexType>>,
+                 public Transposable,
+                 public WritableToMatrixData<ValueType, int32>,
+                 public WritableToMatrixData<ValueType, int64> {
+    friend class Csr<ValueType, IndexType>;
+    friend class EnablePolymorphicObject<Diagonal, LinOp>;
+    friend class EnableCreateMethod<Diagonal>;
+
+public:
+    using value_type = ValueType;
+    using index_type = IndexType;
+    using mat_data = gko::matrix_data<ValueType, int64>;
```

Comment:
I think `mat_data64` would be better

Reply:
I'd prefer leaving it as it is as this is consistent what is done in Dense